### PR TITLE
Implement SurfRide Draw Project

### PR DESCRIPTION
**Pull request by DeaTh-G**
Implement reading and writing of SurfRide Draw Project files into the library.
Tests have been conducted on SWIFv3, SWIFv5 and SWIFv6.

- SurfRide Draw Project (*.swif)
     - [x] Read
     - [x] Write

### CODE REVIEW

File: Source/SharpNeedle/SurfRide/Draw/Animation/AnimationData.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        writer.Write(ID);
+        writer.Write(Count);
+        writer.Write(FrameCount);
+        if (options.Version >= 3)
+            writer.Align(8);
+        if (Count != 0)
+            writer.WriteObjectCollectionOffset(options, this);
+        else
+            writer.WriteOffsetValue(0);
+        if (options.Version >= 1)
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Animation/AnimationData.cs:54]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Animation/AnimationLink.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    public ushort CastID { get; set; }
+    public List<AnimationTrack> Tracks { get; set; } = new();
+
+    public void Read(BinaryObjectReader reader, ChunkBinaryOptions options)
+    {
+        CastID = reader.Read<ushort>();
+        var trackCount = reader.Read<ushort>();
+        if (options.Version >= 3)
+            reader.Align(8);
+        Tracks.AddRange(reader.ReadObjectArrayOffset<AnimationTrack, ChunkBinaryOptions>(options, trackCount));
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Animation/AnimationLink.cs:14]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Animation/AnimationLink.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    public void Write(BinaryObjectWriter writer, ChunkBinaryOptions options)
+    {
+        writer.Write(CastID);
+        writer.Write((ushort)Tracks.Count);
+        if (options.Version >= 3)
+            writer.Align(8);
+        writer.WriteObjectCollectionOffset(options, Tracks);
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Animation/AnimationLink.cs:23]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Animation/AnimationTrack.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        Clear();
+        Field00 = reader.Read<ushort>();
+        Capacity = reader.Read<ushort>();
+        Flags = reader.Read<uint>();
+        StartFrame = reader.Read<uint>();
+        EndFrame = reader.Read<uint>();
+        if (options.Version >= 3)
+            reader.Align(8);
+        switch (Flags & 3)
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Animation/AnimationTrack.cs:20]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Animation/AnimationTrack.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    public void Write(BinaryObjectWriter writer, ChunkBinaryOptions options)
+    {
+        writer.Write(Field00);
+        writer.Write((ushort)Count);
+        writer.Write(Flags);
+        writer.Write(StartFrame);
+        writer.Write(EndFrame);
+        if (options.Version >= 3)
+            writer.Align(8);
+        writer.WriteObjectCollectionOffset(Flags, this);
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Animation/AnimationTrack.cs:48]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Animation/IKeyFrame.cs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            case 0x70:
+                writer.Write(Double);
+                break;
+            case 0x80:
+                writer.Write(Character);
+                break;
+            default:
+                throw new NotImplementedException();
+        }
+        writer.Align(4);
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Animation/IKeyFrame.cs:117]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Animation/IKeyFrame.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            case 0x70:
+                Double = reader.Read<double>();
+                break;
+            case 0x80:
+                Character = reader.Read<char>();
+                break;
+            default:
+                throw new NotImplementedException();
+        }
+        reader.Align(4);
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Animation/IKeyFrame.cs:83]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Camera.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    public Vector3 LookAt { get; set; }
+    public int Field38 { get; set; }
+
+    public void Read(BinaryObjectReader reader, ChunkBinaryOptions options)
+    {
+        if (options.Version >= 1)
+        {
+            if (options.Version >= 3)
+                reader.Align(8);
+            Name = reader.ReadStringOffset();
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Camera.cs:22]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Camera.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+            ID = reader.Read<int>();
+        }
+        if (options.Version >= 4)
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Camera.cs:25]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Camera.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            LookAt = reader.Read<Vector3>();
+            reader.Align(16);
+            Field38 = reader.Read<int>();
+        }
+        else
+        {
+            Position = reader.Read<Vector3>();
+            LookAt = reader.Read<Vector3>();
+        }
+        Flags = reader.Read<uint>();
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Camera.cs:39]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Camera.cs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    public void Write(BinaryObjectWriter writer, ChunkBinaryOptions options)
+    {
+        if (options.Version >= 1)
+        {
+            if (options.Version >= 3)
+                writer.Align(8);
+            writer.WriteStringOffset(StringBinaryFormat.NullTerminated, Name);
+            writer.Write(ID);
+        }
+        if (options.Version >= 4)
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Camera.cs:58]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Camera.cs [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            writer.Write(LookAt);
+            writer.Align(16);
+            writer.Write(Field38);
+        }
+        else
+        {
+            writer.Write(Position);
+            writer.Write(LookAt);
+        }
+        writer.Write(Flags);
```
Comment (Sajidur78) [inline Source/SharpNeedle/SurfRide/Draw/Camera.cs:72]: Missing empty line for formatting.

File: Source/SharpNeedle/SurfRide/Draw/Font.cs [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        writer.WriteObjectCollectionOffset(this);
+        writer.WriteOffsetValue(Field20);
+        if (UserData != null)
+            writer.WriteObjectOffset(UserData, options);
+        else
+            writer.WriteOffsetValue(0);
+    }
+}
+
+public class CharacterMapping : IBinarySerializable
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Font.cs]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Animation/AnimationData.cs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        Clear();
+        if (options.Version >= 3)
+            reader.Align(8);
+        Name = reader.ReadStringOffset();
+        ID = reader.Read<int>();
+        Capacity = reader.Read<int>();
+        FrameCount = reader.Read<uint>();
+        if (options.Version >= 3)
+            reader.Align(8);
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Animation/AnimationData.cs:22]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Animation/AnimationData.cs [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        if (options.Version >= 3)
+            writer.Align(8);
+        writer.WriteStringOffset(StringBinaryFormat.NullTerminated, Name);
+        writer.Write(ID);
+        writer.Write(Count);
+        writer.Write(FrameCount);
+        if (options.Version >= 3)
+            writer.Align(8);
+        if (Count != 0)
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Animation/AnimationData.cs:54]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Animation/AnimationLink.cs (hunk 2 above)
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Animation/AnimationLink.cs:15]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Animation/AnimationTrack.cs (hunk 4 above)
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Animation/AnimationTrack.cs:21]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Animation/AnimationData.cs [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+                    reader.Align(4);
+            }
+        }
+    }
+
+    public void Write(BinaryObjectWriter writer, ChunkBinaryOptions options)
+    {
+        if (options.Version >= 3)
+            writer.Align(8);
+        writer.WriteStringOffset(StringBinaryFormat.NullTerminated, Name);
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Animation/AnimationData.cs:47]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Animation/IKeyFrame.cs (hunk 6 above)
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Animation/IKeyFrame.cs:117]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Camera.cs (hunk 8 above)
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Camera.cs:23]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Camera.cs (hunk 9 above)
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Camera.cs:27]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Camera.cs (hunk 11 above)
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Camera.cs:62]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/CastNode.cs [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+        long userDataOffset = reader.ReadOffsetValue();
+        if (userDataOffset != 0)
+            UserData = reader.ReadObjectAtOffset<UserData, ChunkBinaryOptions>(userDataOffset, options);
+    }
+
+    public void Write(BinaryObjectWriter writer, ChunkBinaryOptions options)
+    {
+        if (options.Version >= 3)
+            writer.Align(8);
+        writer.WriteStringOffset(StringBinaryFormat.NullTerminated, Name);
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/CastNode.cs:48]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/CastNode.cs [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        writer.Write(Flags);
+        if (Cast != null)
+            writer.WriteObjectOffset(Cast, options);
+        else
+            writer.WriteOffsetValue(0);
+        writer.Write(ChildIndex);
+        writer.Write(SiblingIndex);
+        if (options.Version >= 3)
+            writer.Align(8);
+        if (UserData != null)
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/CastNode.cs:61]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/CastNode.cs [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+        if (options.Version >= 3)
+            writer.Align(8);
+        writer.WriteStringOffset(StringBinaryFormat.NullTerminated, Name);
+        writer.Write(ID);
+        writer.Write(Flags);
+        if (Cast != null)
+            writer.WriteObjectOffset(Cast, options);
+        else
+            writer.WriteOffsetValue(0);
+        writer.Write(ChildIndex);
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/CastNode.cs:56]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Font.cs [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+    public ushort Field14 { get; set; }
+    public long Field20 { get; set; }
+    public UserData UserData { get; set; }
+
+    public void Read(BinaryObjectReader reader, ChunkBinaryOptions options)
+    {
+        Clear();
+        if (options.Version >= 3)
+            reader.Align(8);
+        Name = reader.ReadStringOffset();
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Font.cs:19]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Font.cs [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+        Field0C = reader.Read<uint>();
+        if (options.Version >= 4)
+            Field10 = reader.Read<short>();
+        Capacity = reader.Read<ushort>();
+        Field14 = reader.Read<ushort>();
+        if (options.Version >= 3)
+            reader.Align(8);
+        else
+            reader.Align(4);
+        AddRange(reader.ReadObjectArrayOffset<CharacterMapping>(Capacity));
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Font.cs]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Font.cs [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+        long userDataOffset = reader.ReadOffsetValue();
+        if (userDataOffset != 0)
+            UserData = reader.ReadObjectAtOffset<UserData, ChunkBinaryOptions>(userDataOffset, options);
+    }
+
+    public void Write(BinaryObjectWriter writer, ChunkBinaryOptions options)
+    {
+        if (options.Version >= 3)
+            writer.Align(8);
+        writer.WriteStringOffset(StringBinaryFormat.NullTerminated, Name);
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Font.cs:44]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Font.cs [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+        writer.Write(Field0C);
+        if (options.Version >= 4)
+            writer.Write(Field10);
+        writer.Write((ushort)Count);
+        writer.Write(Field14);
+        if (options.Version >= 3)
+            writer.Align(8);
+        else
+            writer.Align(4);
+        writer.WriteObjectCollectionOffset(this);
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Font.cs]: Fixed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/ICast.cs [hunk 24]
```diff
@@ -0,0 +1,24 @@
+﻿namespace SharpNeedle.SurfRide.Draw;
+
+public enum FlagType : uint
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/ICast.cs]: Changed in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/ICell.cs [hunk 25]
```diff
@@ -0,0 +1,15 @@
+﻿namespace SharpNeedle.SurfRide.Draw;
+
+public interface ICastCell : IBinarySerializable<ChunkBinaryOptions>
+{
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/ICell.cs:4]: Setters added and ICastCell renamed to ICell to match filename in commit 7f9fe92.

File: Source/SharpNeedle/SurfRide/Draw/Animation/IKeyFrame.cs (hunk 6 above)
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Animation/IKeyFrame.cs:119]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/Animation/IKeyFrame.cs (hunk 7 above)
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Animation/IKeyFrame.cs:84]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/Camera.cs (hunk 11 above)
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Camera.cs:63]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/ImageCast.cs [hunk 26]
```diff
@@ -0,0 +1,145 @@
+﻿namespace SharpNeedle.SurfRide.Draw;
+
+public class ImageCast : ICast
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/ImageCast.cs:3]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/Layer.cs [hunk 27]
```diff
@@ -0,0 +1,95 @@
+﻿namespace SharpNeedle.SurfRide.Draw;
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Layer.cs:1]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/ReferenceCast.cs [hunk 28]
```diff
@@ -0,0 +1,32 @@
+﻿namespace SharpNeedle.SurfRide.Draw;
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/ReferenceCast.cs:1]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/Scene.cs [hunk 29]
```diff
@@ -0,0 +1,90 @@
+﻿namespace SharpNeedle.SurfRide.Draw;
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Scene.cs:1]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/SliceCast.cs [hunk 30]
```diff
@@ -0,0 +1,185 @@
+﻿namespace SharpNeedle.SurfRide.Draw;
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/SliceCast.cs:1]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/SrdProject.cs [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+                case ProjectChunk project:
+                    Project = project;
+                    break;
+                case TextureListChunk tl:
+                    TextureLists = tl;
+                    break;
+            }
+        }
+        RevisionDate = info.RevisionDate;
+
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/SrdProject.cs:45]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/Texture.cs [hunk 32]
```diff
@@ -0,0 +1,61 @@
+﻿namespace SharpNeedle.SurfRide.Draw;
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/Texture.cs:1]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/TextureListChunk.cs [hunk 33]
```diff
@@ -0,0 +1,88 @@
...
+
+        var size = (long)end - (long)start;
+        writer.At((long)start - sizeof(int), SeekOrigin.Begin);
+        writer.Write((int)size);
+
+        end.Dispose();
+    }
+}
+
+public class TextureList : List<Texture>, IBinarySerializable<ChunkBinaryOptions>
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/TextureListChunk.cs:43]: Fixed in commit 534bef1.

File: Source/SharpNeedle/SurfRide/Draw/UserData.cs [hunk 34]
```diff
@@ -0,0 +1,128 @@
+﻿namespace SharpNeedle.SurfRide.Draw;
```
[the remark this answers is not in the document]
Comment (DeaTh-G) [reply Source/SharpNeedle/SurfRide/Draw/UserData.cs:1]: Fixed in commit 534bef1.

**Pull request merged**